src/platform/windows/nvprefs/undo_file.cpp

```diff
@@ -0,0 +1,154 @@
+#include "nvprefs_common.h"
+
+#include "undo_file.h"
+
+namespace {
+
+  using namespace nvprefs;
+
+  DWORD
+  relax_permissions(HANDLE file_handle) {
+    PACL old_dacl = nullptr;
+
+    safe_hlocal<PSECURITY_DESCRIPTOR> sd;
+    DWORD status = GetSecurityInfo(file_handle, SE_FILE_OBJECT, DACL_SECURITY_INFORMATION, nullptr, nullptr, &old_dacl, nullptr, &sd);
+    if (status != ERROR_SUCCESS) return status;
+
+    safe_sid users_sid;
+    SID_IDENTIFIER_AUTHORITY nt_authorithy = SECURITY_NT_AUTHORITY;
+    if (!AllocateAndInitializeSid(&nt_authorithy, 2, SECURITY_BUILTIN_DOMAIN_RID, DOMAIN_ALIAS_RID_USERS, 0, 0, 0, 0, 0, 0, &users_sid)) {
+      return GetLastError();
+    }
+
+    EXPLICIT_ACCESS ea = {};
+    ea.grfAccessPermissions = GENERIC_READ | GENERIC_WRITE | DELETE;
+    ea.grfAccessMode = GRANT_ACCESS;
+    ea.grfInheritance = NO_INHERITANCE;
+    ea.Trustee.TrusteeForm = TRUSTEE_IS_SID;
+    ea.Trustee.ptstrName = (LPTSTR) users_sid.get();
+
+    safe_hlocal<PACL> new_dacl;
+    status = SetEntriesInAcl(1, &ea, old_dacl, &new_dacl);
+    if (status != ERROR_SUCCESS) return status;
+
+    status = SetSecurityInfo(file_handle, SE_FILE_OBJECT, DACL_SECURITY_INFORMATION, nullptr, nullptr, new_dacl.get(), nullptr);
+    if (status != ERROR_SUCCESS) return status;
+
+    return 0;
+  }
+
+}  // namespace
+
+namespace nvprefs {
+
+  std::optional<undo_file_t>
+  undo_file_t::open_existing_file(std::filesystem::path file_path, bool &access_denied) {
+    undo_file_t file;
+    file.file_handle.reset(CreateFileW(file_path.c_str(), GENERIC_READ | DELETE, 0, nullptr, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL));
+    if (file.file_handle) {
+      access_denied = false;
+      return file;
+    }
+    else {
+      auto last_error = GetLastError();
+      access_denied = (last_error != ERROR_FILE_NOT_FOUND && last_error != ERROR_PATH_NOT_FOUND);
+      return std::nullopt;
+    }
+  }
+
+  std::optional<undo_file_t>
+  undo_file_t::create_new_file(std::filesystem::path file_path) {
+    undo_file_t file;
+    file.file_handle.reset(CreateFileW(file_path.c_str(), GENERIC_WRITE | STANDARD_RIGHTS_ALL, 0, nullptr, CREATE_NEW, FILE_ATTRIBUTE_NORMAL, NULL));
+
+    if (file.file_handle) {
+      // give GENERIC_READ, GENERIC_WRITE and DELETE permissions to Users group
+      if (relax_permissions(file.file_handle.get()) != 0) {
```

Comment:
What's the thinking here for allowing the Users group to write this file (or really for any user/process besides us to write it)?

Reply:
It's really GENERIC_READ and DELETE that's is required logic-wise - read file, apply undo, delete file. But GENERIC_WRITE is needed for locking the file during the process.

"Us" is defined here by file path, not user or process. Since we have portable version that can be be run in parallel and as any user, and then can crash and leave the undo file with settings not reverted. If NVIDIA state is global and can be changed by anyone, then the undo file must be global too, I don't think there's a way around it.